utils/near-clock/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+use std::cell::RefCell;
+use std::ops::{Add, Mul, Sub};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::{Duration, SystemTime};
+
+thread_local! {
+    // A shared pointer. We should propagate it across multiple threads.
+    // Replace by AtomicU128 by 2262-04-12, after that date 64 bits won't be enough to
+    // represent time in nanoseconds. AtomicU128 is still in nightly build only.
+    static TIME_OVERRIDE: RefCell<Arc<AtomicU64>> = RefCell::new(Arc::new(AtomicU64::new(0)));
+}
+
+// Used for tests to set clock time manually.
+pub fn mock_time(new_time: NearClock) {
+    TIME_OVERRIDE.with(|time_override| {
+        let ns = new_time.duration_since(&UNIX_EPOCH);
+        time_override.borrow().store(ns.as_nanos() as u64, Ordering::SeqCst)
+    });
+}
+
+pub const UNIX_EPOCH: NearClock = NearClock::from_system_time(SystemTime::UNIX_EPOCH);
+
+// A pointer to shared atomic, which is used to override time.
+// This pointer is meant to be shared between threads. A typical usage would be to get pointer
+// to atomic on the main thread running unit test, and then propagate is to sub threads
+pub fn get_time_override() -> Arc<AtomicU64> {
+    TIME_OVERRIDE.with(|x| x.borrow().clone())
+}
+
+// sets override on child threads
+pub fn set_time_override(new_time_override: Arc<AtomicU64>) {
+    TIME_OVERRIDE.with(|x| *x.borrow_mut() = new_time_override);
+}
+
+// get current time; we should use it everywhere in our code base
+pub fn now() -> NearClock {
+    TIME_OVERRIDE.with(|time_override| {
+        let time_override = time_override.borrow().load(Ordering::SeqCst);
+        if time_override == 0 {
+            NearClock { time: SystemTime::now() }
+        } else {
+            NearClock { time: SystemTime::UNIX_EPOCH + Duration::from_nanos(time_override) }
+        }
+    })
+}
+
+// TODO: Add other methods that `SystemTime` has.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub struct NearClock {
+    time: SystemTime,
+}
+
+impl NearClock {
+    pub const fn from_system_time(system_time: SystemTime) -> NearClock {
+        NearClock { time: system_time }
+    }
+
+    pub fn now() -> NearClock {
+        now()
+    }
+
+    pub fn duration_since(&self, rhs: &NearClock) -> Duration {
+        self.time.duration_since(rhs.time).unwrap_or(Duration::from_millis(0))
```

Comment:
```suggestion
        self.time.duration_since(rhs.time).unwrap_or_default()
```

A bit less explicit, but more idiomatic and minimal. For pure Rust code-bases, I think "use the most direct solution within the language" is a useful consistency guideline.